python/amici/sbml_import.py

```diff
@@ -1321,10 +1337,40 @@ def _add_conservation_for_constant_species(
                 # mark species to delete from stoichiometrix matrix
                 species_solver.pop(ix)
 
+        return species_solver
+
+    def _reduce_stoichiometry(self, species_solver, volume_updates) -> List:
+        """
+        Reduces the stoichiometry with respect to conserved quantities
+
+        :param species_solver:
+            List of species indices which remain later in the ODE solver
+
+        :param volume_updates:
+            List with updates for the stoichiometrix matrix accounting for
+            compartment volumes
+
+        :returns volume_updates_solver:
+            List (according to reduced stoichiometry) with updates for the
+            stoichiometrix matrix accounting for compartment volumes
+        """
+
         # prune out constant species from stoichiometric matrix
         self.stoichiometric_matrix = \
             self.stoichiometric_matrix[species_solver, :]
 
+        # updates of stoichiometry (later dxdotdw in ode_exporter) must be
+        # corrected for conserved quantities:
+        volume_updates_solver = []
+        for update in volume_updates:
+            x_index = update[0]
+            if x_index in species_solver:
+                x_index = species_solver.index(x_index)
```

Comment:
overwriting the irator in a loop is a recipe for confusion.

Reply:
I had that changed at some point. You're definitely right!
Will change that...
How/when best? We will have another PR very soon anyhow...

Reply:
especially since thats also the iterator of the outer loop, maybe I'm missing some wild recursion here, but I don't think this function is doing what its supposed to do.

Reply:
Ah, right. No, that's actually fine. `update` is immutable and a tuple...
Hence, line 1366 creates a new obejct and doesn't overwrite ;)

Reply:
This makes it safe, but still bad practice. Will change that.

Reply:
the logic in this function is so convoluted, I can't really say what its supposed to do. According to the docstring, x_index is already a list of indices, so I don't understand why you have to index over that again.

From a design perspective, its probably more robust to generally work with identifiers here instead of indices as we can reconstruct the index ad hoc when the update is applied.

Reply:
> Ah, right. No, that's actually fine. `update` is immutable and a tuple...
> Hence, line 1366 creates a new obejct and doesn't overwrite ;)

Yeah, overwriting the original value would be even worse, but currently the value of update[0] goes completely unused. If that is intended, then assign update[0] to `_` in the first place. Like this it looks like something is fishy in the implementation.

Reply:
Need better doc then...
No, x_index is an int. Actually, this thing is supposed to be rather simple. It's one loop and one if, in which the first index is "corrected"...

Reply:
So you are trying to do this:
```
for _, x, y in volume_updates:
    volume_updates_solver.extend([(ix, x, y) for ix in range(len(species_solver))])
```
?

Its not prohibited to iterate over indices in python, I would actually prefer that over iterating over elements and then looking up the indices.

Reply:
ah, wait, now I see more doc is needed...
This indexing is not just for looking up! It provides the mapping between x_rdata and x_solver.

Reply:
> So you are trying to do this:
> 
> ```
> for _, x, y in volume_updates:
>     volume_updates_solver.extend([(ix, x, y) for ix in range(len(species_solver))])
> ```
> 
> ?
> 
> Its not prohibited to iterate over indices in python, I would actually prefer that over iterating over elements and then looking up the indices.

Yes, this would work, but assume that updates are always passed in the right order...
If there is no way whatsoever that the state ordering is changed by conservation laws, then it's fine...

Reply:
happy to be proven wrong, but I am pretty sure that this function just reimplements what is currently being done. I agree that using indices is not the best way to go about this.

Reply:
you mean the whole reduce_stoichiometry function?
If there is already a function with an API for this, it wasn't used in SBML import so far. SBML import was lacking this reduction until now. But there may be some functionality I am not aware of...

Reply:
```
for update in volume_updates:
    x_index = update[0]
    if x_index in species_solver:
        x_index = species_solver.index(x_index)
        volume_updates_solver.append((x_index, update[1], update[2]))
```
is equivalent to
```
for _, x, y in volume_updates:
    volume_updates_solver.extend([(ix, x, y) for ix in range(len(species_solver))])
```

intermediate steps:
```
for x_index, x, y in volume_updates:
    if x_index in species_solver:
        x_index = species_solver.index(x_index)
        volume_updates_solver.append((x_index, x, y))
```

```
for _, x, y in volume_updates:
    if x_index in species_solver:
        x_index = species_solver.index(x_index)
        volume_updates_solver.append((x_index, x, y))
```

```
for _, x, y in volume_updates:
    if x_index in range(len(species_solver)):
        volume_updates_solver.append((x_index, x, y))
```

not a big fan of having hour long conversations about four lines of code ...